### ISSUE TITLE
Switch to fuse 3.x

### ISSUE THREAD

**Issue by Keruspe** [s3fs-fuse/s3fs-fuse#1159]
From the release notes of libfuse 2.9.9:

```
Note: This is a maintenance release from the 2.9 branch of libfuse. Users are encouraged to transition to the actively developed libfuse 3.x.
```

**Comment by gaul** [comment]
Can you share what advantages this would provide?  While CentOS 7 EPEL includes fuse3, only Ubuntu 19.04 includes fuse3 which means that both 16.04 and 18.04 LTS users would not be able to use s3fs with fuse3.

**Comment by juliogonzalez** [comment]
CentOS6 will break (but EoL is not that far away).

It would also break:
- SLE12 (all SPs)
- SLE15 (but not SLE15SP1)
- openSUSE Leap 15.0 (but not Leap 15.1)

All **maintained** Fedora versions seem to be have already 3.X

**Comment by Keruspe** [comment]
Most obvious advantage would be to depend on a maintained version of fuse instead of a soon-to-be-deprecated one

**Comment by gaul** [comment]
If there are no functional benefits then I do not see a reason to pursue this.  After Ubuntu 18.04 ends support in 2023 we can re-evaluate this.

**Comment by juliogonzalez** [comment]
As far as I can see at `configure.ac`, s3fs-fuse requires use `2.8.4` as **minimum**

@Keruspe does master compile with 3.X?

If it does, then there's nothing to do at this moment.

If it doesn't, then it's not critical, but at some point (most probably before 2023), this should be addressed, as I am pretty sure newer distributions will start deprecating removing `2.x`, while older distributions will keep it.

**Comment by ggtakec** [comment]
@Keruspe Changes from libfuse 2.x to 3.x include the addition of arguments to some I / F functions.
Therefore, it involves a modification of the s3fs code.
I want these source codes to be compatible with each other (not #ifdef), and to make s3fs source code be one codes.
We will make s3fs correspond to libfuse 3.x, but please take some time including testing and investigation.
Thanks in advance for your kindness.

**Comment by gaul** [comment]
It appears that FUSE 3.6.0 supports 1 MB writes:

https://github.com/libfuse/libfuse/blob/master/ChangeLog.rst#libfuse-360-2019-06-13

Previously s3fs saw a large benefit when moving from 4 KB to 128 KB writes #813.

**Comment by gaul** [comment]
I have done as much as I can do with #1867 -- modern Linux is ready for FUSE 3.  I don't use macOS and don't understand its FUSE situation but this seems like the biggest blocker.  I would appreciate help from the community.

**Comment by ggtakec** [comment]
I think it is possible to support both FUSE2 and FUSE3 with the same source code(using ifdef etc.).
This is a suggestion, but is it possible to separate the packages and support them?(It's not a good name, but s3fs-fuse3, etc.)
If the source code can support both, I think it can only be a matter of how to use the provided package.
Thus, is it possible to omit this issue blocker by suspending the provision of the FUSE3 package on an OS that is difficult to implement?

**Comment by gaul** [comment]
@breisig Which Linux distribution does not provide FUSE2?

**Comment by jrandall** [comment]
This is also an issue for singularity --fusemount support, which only works with filesystems built with libfuse3:
```
The FUSE command must be based on libfuse3 to work correctly with Singularity --fusemount. If you are using an older distribution that provides FUSE commands such as sshfs based on FUSE 2 then you can install FUSE 3 versions of the commands you need inside your container.
```
(from https://docs.sylabs.io/guides/3.6/user-guide/bind_paths_and_mounts.html#fuse-mounts)

s3fs-fuse does not currently work with singularity because of this.

I believe the reason that fuse 3 is required is that they are using functionality added in libfuse 3.3.0 that allows a privileged process (such as singularity) to mount `/dev/fuse` and then pass it as a special file descriptor to the filesystem (in place of a mount point): https://github.com/libfuse/libfuse/blob/master/ChangeLog.rst#libfuse-330-2018-11-06

**Comment by ggtakec** [comment]
@jrandall 
I am interested in the following features:
https://github.com/libfuse/libfuse/blob/master/ChangeLog.rst#libfuse-330-2018-11-06
And I think it affects #1246 as well.
This issue is closed, but we will continue to consider support for FUSE3.
_I have other work to do, so I can't do it right away, but I'm going to proceed with the investigation._

Thank you for your valuable opinion.

**Comment by adamqqqplay** [comment]
> It appears that FUSE 3.6.0 supports 1 MB writes:
> 
> https://github.com/libfuse/libfuse/blob/master/ChangeLog.rst#libfuse-360-2019-06-13
> 
> Previously s3fs saw a large benefit when moving from 4 KB to 128 KB writes #813.

@gaul Yeah, we are also very interested in this feature, it should improve the performance of sequential IO scenarios.

**Comment by gaul** [comment]
The current macOS FUSE situation is confusing but I don't believe it supports FUSE 3 yet: https://github.com/osxfuse/osxfuse/issues/390

**Comment by ggtakec** [comment]
This needs further investigation, but from what I've seen before, the differences between fuse2 and fuse3 interface can be supported with the same source code.
About interface, I beleive that s3fs has `pseudo fd` prepared internally, and by making good use of this, I think the differences between fuse2 and fuse3 can be reduced.
And I think it looks like statsx support can be added to the existing stat information. (FUSE_DIRECT_IO_RELAX has not been confirmed.)
However, this will require the introduction of a few `ifdef`, or dynamic loading of the fuse library.

**Comment by gaul** [comment]
I don't think s3fs should take on FUSE2+FUSE3 complexity since all the supported Linux distributions support FUSE3.  If the FUSE3 complexity is small, then FUSE-T should make small changes to support their users to enable FUSE3.  If it is large then I don't think it is worth our effort.  FUSE2 has not had a release since 2008!

**Comment by ggtakec** [comment]
> I don't think s3fs should take on FUSE2+FUSE3 complexity since all the supported Linux distributions support FUSE3. 

Thank you. This was something I hadn't checked properly.
If there are no problems with installing FUSE3 on each distribution, I don't think there is any need to keep FUSE2 compatibility in s3fs.

**Comment by juliogonzalez** [comment]
About this, not saying this should not be done, but be aware CentOS7 is not the only OS that will go away (CentOS7 should not matter much as it's EoL, even if some clones are not).

Some other OS without FUSE3 (at least not as an official package):
- Amazon Linux 2: EoL is 2025/06/30
- SUSE Linux Enterprise 12 (all service packs): EoL is 2030/10/31
- Ubuntu 20.04: EoL is April 2025 (extra paid support of course will last until 2030/2032)
  FUSE3 is available in Universe, but remember universe is not supported, although is a workaround.

Of course for RPMs providing FUSE3 could be an alternative to be considered if needed and if it's not too complicated

Every other recent distribution I could check has FUSE3, namely:
- SUSE Linux Enterprise 15
- openSUSE Leap 15 and openSUSE Tumbleweed
- AlmaLinux 8 and newer
- Amazon Linux 2023
- Ubuntu 22.04 and newer
- Debian 10 and newer

So in short, the bump to FUSE3 can be done, but if done before summer 2025, I'd recommend mentioning at the release note the OS that won't be supported.

Related issues: https://github.com/s3fs-fuse/s3fs-fuse/issues/2469 and https://github.com/s3fs-fuse/s3fs-fuse/issues/2496

**Comment by ggtakec** [comment]
I think that modifying the code to support FUSE3(as an initial stage) will not be too difficult, and I am ready to get started.
And at that initial stage, I think the code for FUSE3 and FUSE2 will be able to coexist.
(I think they can be separated by configure or build flags.)

@gaul
Once I've finished my current work, I'll try preparing the code.
(Basically, it's a change to the interface of the functions registered in FUSE, but that can be handled by using pseudofd.)

**Comment by gaul** [comment]
Debian requests that s3fs moves to FUSE 3: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1084409

**Comment by rjschwei** [comment]
And now there is a bug report for openSUSE as well. https://bugzilla.suse.com/show_bug.cgi?id=1242091

**Comment by gaul** [comment]
I prefer to just require FUSE 3.0 and drop 2.x support.  While supporting both is possible I think packagers can just use an older version.

**Comment by andreasstieger** [comment]
Correct, the situation in the distributions is that fuse3 is available everywhere, and they are trying to get off fuse 2 altogether.

**Comment by gaul** [comment]
I hope someone can contribute benchmarks for this.
